### PR TITLE
postgres: remove public schema creation for hive database

### LINKED ISSUE #28: Hive (metastore) uses ```public``` Postgres schema to store SQL tables
According to the script SQL ```hive-schema-3.1.0.postgres.sql``` used by ```hive``` to initiate schema in Postgres database, it seems that ```hive``` forces to use the ```public``` schema to create SQL tables, this is hard coded in the script, line 12:

```SET search_path = public, pg_catalog;```

Thus in ```postgresql``` role, there will not need to create a SQL schema in ```hive``` database.

### PR DESCRIPTION
**Pull request by kpgtek**
#### Which issue(s) this PR fixes:

Fixes #28

#### Additional comments:

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0) and that you give permission to [TOSIT](https://www.tosit.io/),
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] I hereby agree to grant [TOSIT](https://www.tosit.io/) a copyright license to use my contributions.

### CODE REVIEW

File: roles/postgresql/defaults/main.yml [hunk 1]
```diff
@@ -126,10 +126,13 @@ postgresql_additional_databases: []
...
 postgresql_databases: "{{ postgresql_generated_databases + postgresql_additional_databases }}"
 
+#
+# Hive metastore uses "public" schema hardcoded, no schema is created for Hive (issue #28)
+#
 postgresql_generated_schemas: >-
   [
     {%- for db in postgresql_databases -%}
-      {%- if db.owner is defined -%}
+      {%- if db.owner is defined and db.owner != postgresql_hive_user -%}
```
Comment (gboutry) [inline roles/postgresql/defaults/main.yml:135]: We should make this more generic

We should create a list such as `no_custom_schemas_users` and check with:
```suggestion
      {%- if db.owner is defined and not db.owner in no_custom_schemas_users -%}
```

File: roles/postgresql/defaults/main.yml [hunk 2]
```diff
@@ -126,10 +126,13 @@ postgresql_additional_databases: []
 
 postgresql_databases: "{{ postgresql_generated_databases + postgresql_additional_databases }}"
 
+#
+# Hive metastore uses "public" schema hardcoded, no schema is created for Hive (issue #28)
```
Comment (gboutry) [inline roles/postgresql/defaults/main.yml:130]:
```suggestion
# Hive MetaStore use of "public" schema is hardcoded in its scripts, therefore we do not create a specific schema for Hive (issue #28)
```

**Review by gboutry — CHANGES_REQUESTED** [review]

File: roles/postgresql/defaults/main.yml (hunk 1 above)
Comment (rpignolet) [reply roles/postgresql/defaults/main.yml:135]: I propose you add an attribute `schema_generated` to items in `postgresql_databases`. You have an example with `hba_entry_generated` here https://github.com/TOSIT-IO/tdp-collection-prerequisites/blob/466a9d19da2548b31fdbb38139bd5b705aac3444/roles/postgresql/defaults/main.yml#L150-L153

**Review by gboutry — APPROVED** [review]

File: roles/postgresql/defaults/main.yml [hunk 3]
```diff
@@ -75,6 +75,11 @@ postgresql_jupyterhub_user: jupyterhub
 postgresql_jupyterhub_password: jupyterhub
 postgresql_jupyterhub_database: jupyterhub
 
+no_custom_schemas_users: >-
```
Comment (rpignolet) [inline roles/postgresql/defaults/main.yml:78]: I do not understand why you create this variable, you add `schema_generated` to `postgresql_generated_schemas` so why not using it ?

File: roles/postgresql/defaults/main.yml [hunk 4]
```diff
@@ -119,17 +119,22 @@ postgresql_generated_databases: >-
...
 
+#
+# Hive MetaStore use of "public" schema is hardcoded in its scripts, therefore we do not create a specific schema for Hive (issue #28)
+# "schema_generated" is set to "no" in "hive" database
+#
 postgresql_generated_schemas: >-
   [
     {%- for db in postgresql_databases -%}
-      {%- if db.owner is defined -%}
+      {%- if db.owner is defined and (not db.schema_generated is defined or not db.schema_generated|length or db.schema_generated == "yes") -%}
```
Comment (rpignolet) [inline roles/postgresql/defaults/main.yml:137]: Can you test this condition if it works ?
```suggestion
      {%- if db.owner is defined and (db.schema_generated is not defined or db.schema_generated) -%}
```
You can see this pattern here https://github.com/TOSIT-IO/tdp-collection-prerequisites/blob/b2de13e2b70b71e9a98ae8cae1b57c519728d754/roles/postgresql/defaults/main.yml#L157

**Review by rpignolet — APPROVED** [review]

File: roles/postgresql/defaults/main.yml (hunk 4 above)
Comment (kpgtek) [reply roles/postgresql/defaults/main.yml:137]: For ```hive```, if I use the following condition:
```
      {%- if db.owner is defined and (db.schema_generated is not defined or db.schema_generated) -%}
```
- if ```schema_generated``` is not set in database, ```hive``` schema is generated
- if ```"schema_generated": ""``` set to empty in database, ```hive``` schema is not generated
- if ```"schema_generated": "yes"``` is set to "yes" in database, ```hive``` schema is generated
- if ```"schema_generated": "no"``` is set to "no" in database, ```hive``` schema is generated

Comment (rpignolet) [reply roles/postgresql/defaults/main.yml:137]: `yes` and `no` are not a boolean for a JSON, the Jinja code in the variable output a JSON so JSON type should be use and a boolean in JSON is `true` or `false`.

>     * if `schema_generated` is not set in database, `hive` schema is generated
I agree
>     * if `"schema_generated": ''` set to empty in database, `hive` schema is not generated
`schema_generated` is a boolean, `''` is empty string which is not a boolean, you can ignore this case.
>     * if `"schema_generated": true` is set to "True" in database (to replace "yes"), `hive` schema is generated
I agree
>     * if `"schema_generated": false` is set to "False" in database (to replace "no"), `hive` schema is not generated
I agree

So the condition I propose works no ?

Comment (kpgtek) [reply roles/postgresql/defaults/main.yml:137]: Yes, your condition works for the flags ```schema_generated``` and ```hba_entry_generated``` set to boolean.
So I think it is worth to clarify in lines 122 and 123 that the flags are boolean (set to 'true' and not to 'yes'):
```
  #   schema_generated: true # default (auto generate schema list)   
  #   hba_entry_generated: true # default (activate auto generated pg_hba entry)
```

Comment (rpignolet) [reply roles/postgresql/defaults/main.yml:137]: Yes, update the doc to use `true` or `false`.

**Review by rpignolet — APPROVED** [review]

**Pull request merged**